include/targets/cuda/shared_memory_helper.h

```diff
@@ -0,0 +1,89 @@
+#pragma once
+
+#include <target_device.h>
+
+/**
+   @file shared_memory_helper.h
+
+   Target specific helper for allocating and accessing shared memory.
+ */
+
+namespace quda
+{
+
+  /**
+     @brief Class which is used to allocate and access shared memory.
+     The shared memory is treated as an array of type T, with the
+     number of elements given by a call to the static member
+     S::size(target::block_dim()).  The byte offset from the beginning
+     of the total shared memory block is given by the static member
+     O::shared_mem_size(target::block_dim()), or 0 if O is void.
+   */
+  template <typename T, typename S, typename O = void> class SharedMemory
+  {
+  public:
+    using value_type = T;
+
+  private:
+    T *data;
+
+    /**
+       @brief This is a dummy instantiation for the host compiler
+    */
+    template <bool, typename dummy = void> struct cache_dynamic {
+      T *operator()(unsigned int)
+      {
+        static T *cache_;
+        return cache_;
+      }
+    };
+
+    /**
+       @brief This is the handle to the dynamic shared memory
+       @return Shared memory pointer
+     */
+    template <typename dummy> struct cache_dynamic<true, dummy> {
```

Comment:
In this rewrite, we only have dynamic shared memory now for the CUDA target.  Were there any kernels that still use the static shared memory approach?  Just wondering if any .cu files need to be updated, e.g., need  `sharedBytesPerThread()` or `sharedBytesPerBlock()` to be updated?

Reply:
I didn't encounter any static shared memory use, so I don't think anything needs updating.